mlflow/utils/autologging_utils/client.py

```diff
@@ -0,0 +1,370 @@
+            logging_operations.await_completion()
+        return logging_operations
+
+    def _get_pending_operations(self, run_id):
+        """
+        :return: A `_PendingRunOperations` containing all pending operations for the
+                 specified `run_id`.
+        """
+        if run_id not in self._pending_ops_by_run_id:
+            self._pending_ops_by_run_id[run_id] = _PendingRunOperations(run_id=run_id)
+        return self._pending_ops_by_run_id[run_id]
+
+    def _try_operation(self, fn, *args, **kwargs):
+        """
+        Attempt to evaluate the specified function, `fn`, on the specified `*args` and `**kwargs`,
+        returning either the result of the function evaluation (if evaluation was successful) or
+        the exception raised by the function evaluation (if evaluation was unsuccessful).
+        """
+        try:
+            return fn(*args, **kwargs)
+        except Exception as e:
+            return e
+
+    def _flush_pending_operations(self, pending_operations):
+        """
+        Synchronously and sequentially flushes the specified list of pending run operations.
+
+        NB: Operations are not parallelized on a per-run basis because MLflow's File Store, which
+        is frequently used for local ML development, does not support threadsafe metadata logging
+        within a given run.
+        """
+        if pending_operations.create_run:
+            create_run_tags = pending_operations.create_run.tags
+            num_additional_tags_to_include_during_creation = MAX_ENTITIES_PER_BATCH - len(
+                create_run_tags
+            )
+            if num_additional_tags_to_include_during_creation > 0:
+                create_run_tags.extend(
+                    pending_operations.tags_queue[:num_additional_tags_to_include_during_creation]
+                )
+                pending_operations.tags_queue = pending_operations.tags_queue[
+                    num_additional_tags_to_include_during_creation:
+                ]
+
+            new_run = self._client.create_run(
+                experiment_id=pending_operations.create_run.experiment_id,
+                start_time=pending_operations.create_run.start_time,
+                tags={tag.key: tag.value for tag in create_run_tags},
+            )
+            pending_operations.run_id = new_run.info.run_id
+
+        run_id = pending_operations.run_id
+        assert not isinstance(run_id, PendingRunId), "Run ID cannot be pending for logging"
+
+        operation_results = []
+
+        param_batches_to_log = chunk_list(
+            pending_operations.params_queue, chunk_size=MAX_PARAMS_TAGS_PER_BATCH,
+        )
+        tag_batches_to_log = chunk_list(
+            pending_operations.tags_queue, chunk_size=MAX_PARAMS_TAGS_PER_BATCH,
+        )
+        for params_batch, tags_batch in zip_longest(
+            param_batches_to_log, tag_batches_to_log, fillvalue=[]
+        ):
+            metrics_batch_size = min(
```

Comment:
Could this number become negative? It seems not the intention to use it that way below.

Reply:
Good callout! I've added logic to ensure that the value is never negative.